### PR TITLE
Escape character support for string literals

### PR DESCRIPTION
**Pull request by MitchellGale-BitQuill**
### Description
Adds support for escape characters `\` to be used to escape characters in string literals. Keeps support for double quote to act as an escape if the surrounding quote is matching type (single quote, or double quote).

```
SELECT "Hello", 'Hello', "It""s", 'It''s', "It's", '"Its"', 'It\'s', 'It\\\'s', "\I\t\s"
fetched rows / total rows = 1/1
+-----------+-----------+-----------+-----------+----------+-----------+-----------+-------------+------------+
| "Hello"   | 'Hello'   | "It""s"   | 'It''s'   | "It's"   | '"Its"'   | 'It\'s'   | 'It\\\'s'   | "\I\t\s"   |
|-----------+-----------+-----------+-----------+----------+-----------+-----------+-------------+------------|
| Hello     | Hello     | It"s      | It's      | It's     | "Its"     | It's      | It\'s       | \I\t\s     |
+-----------+-----------+-----------+-----------+----------+-----------+-----------+-------------+------------+
```

### Issues Resolved
https://github.com/opensearch-project/sql/issues/297
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -8,7 +8,7 @@ buildscript {
     ext {
         opensearch_version = System.getProperty("opensearch.version", "3.0.0-SNAPSHOT")
         spring_version = "5.3.22"
-        jackson_version = "2.13.4"
+        jackson_version = "2.14.0"
```
Comment (acarbonetto) [inline build.gradle:11]: I don't think we should be checking this in.  It's unrelated content :)

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java [hunk 2]
```diff
@@ -67,13 +48,18 @@ public static String unquoteText(String text) {
     for (int chIndex = 1; chIndex < text.length() - 1; chIndex++) {
       currentChar = text.charAt(chIndex);
       nextChar = text.charAt(chIndex + 1);
-      if (currentChar == enclosingQuote
-          && nextChar == currentChar) {
+
+      if ((currentChar == '\\'
```
Comment (acarbonetto) [inline common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:52]: logic here is very hard to follow (containing both && and || and lots of brackets. I'd suggest simplifying it. 

May have one if statement for `currentChar == '\\'` and another for `currentChar == enclosingQuote`

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java [hunk 3]
```diff
@@ -12,52 +12,33 @@
...
    *     removed
    */
   public static String unquoteText(String text) {
-
     if (text.length() < 2) {
       return text;
     }
 
-    char enclosingQuote;
+    char enclosingQuote = 0;
```
Comment (acarbonetto) [inline common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:30]: can't we default it to firstChar?

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java [hunk 4]
```diff
@@ -12,52 +12,33 @@
...
     char firstChar = text.charAt(0);
     char lastChar = text.charAt(text.length() - 1);
```
Comment (acarbonetto) [inline common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:32]: nit: Suggest renaming to:  `closingQuote`

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java [hunk 5]
```diff
@@ -12,52 +12,33 @@
...
    */
   public static String unquoteText(String text) {
-
     if (text.length() < 2) {
       return text;
     }
 
-    char enclosingQuote;
+    char enclosingQuote = 0;
     char firstChar = text.charAt(0);
```
Comment (acarbonetto) [inline common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:31]: suggest removing `firstChar` and just using `enclosingQuote`

File: integ-test/src/test/java/org/opensearch/sql/sql/StringLiteralIT.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+import static org.opensearch.sql.util.MatcherUtils.rows;
+import static org.opensearch.sql.util.MatcherUtils.schema;
+import static org.opensearch.sql.util.MatcherUtils.verifyDataRows;
+import static org.opensearch.sql.util.MatcherUtils.verifySchema;
+
+
+
+public class StringLiteralIT extends SQLIntegTestCase {
+//  @Test
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/StringLiteralIT.java:23]: TODO

**Review by acarbonetto — CHANGES_REQUESTED** [review]

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java (hunk 3 above)
Comment (MitchellGale-BitQuill) [reply common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:30]: Dangerous if there is no surrounding quotes to a string passed.

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java (hunk 4 above)
Comment (MitchellGale-BitQuill) [reply common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:32]: It's both the opening and closing quote, if the closing quote != opening quote it will be confusing if it's named `closingQuote`.

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java (hunk 5 above)
Comment (MitchellGale-BitQuill) [reply common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:31]: I don't like setting enclosing to something like that in the case where the starting quote != ending quote. Could cause a problem later.

File: build.gradle (hunk 1 above)
Comment (MitchellGale-BitQuill) [reply build.gradle:11]: Build/GHA fail without the change. Will be rebased when https://github.com/opensearch-project/sql/pull/1066 merges.

**Review by margarit-h — APPROVED** [review]

File: common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java (hunk 2 above)
Comment (acarbonetto) [reply common/src/main/java/org/opensearch/sql/common/utils/StringUtils.java:52]: Definitely need to put a set of brackets around 
```
currentChar == nextChar
          && currentChar == enclosingQuote
```

**Review by acarbonetto — APPROVED** [review]

**Comment by GumpacG** [comment]
Does this make that test in TDVT pass? Might be good to check as that is the issue being solved by this PR.

**Review by GumpacG — APPROVED** [review]

**Review by MaxKsyunz — APPROVED** [review]
